src/sinks/util/buffer/compression.rs

```diff
@@ -99,9 +100,34 @@ impl<'de> de::Deserialize<'de> for Compression {
                             if level.is_some() {
                                 return Err(de::Error::duplicate_field("level"));
                             }
-                            level = Some(match map.next_value::<usize>() {
-                                Ok(value) => value.to_string(),
-                                Err(_) => map.next_value::<&str>()?.to_string(),
+                            level = Some(match map.next_value::<Value>()? {
+                                Value::Number(level) => match level.as_u64() {
+                                    Some(value) if value <= 9 => value as usize,
+                                    Some(_) | None => {
+                                        return Err(de::Error::invalid_value(
+                                            de::Unexpected::Other("0, 1, 2, 3, 4, 5, 6, 7, 8 or 9"),
```

Comment:
First parameter of `invalid_value` is the invalid value in the input, so we need to pass `level` here. I'm not sure what to do with the `None` case, maybe we convert it to a string if it is too cumbersome to conver `serde_json::Number` into `Unexpected`. 
 
Second parameter is what data was expected, using `self` as the second parameter will print "expected string or map". You can use a reference to a string here, so `&"0, 1, 2, 3, 4, 5, 6, 7, 8 or 9"` would be perfect.